File: python/question1/question1.py
#!/usr/bin/env python

import fileinput
import sys

class circular_buffer:
    def __init__(self, size):
        self.buffer = [None] * size
        self.head = 0
        self.tail = 0
        self.lastWasAdd = False

    def add(self, value):
        self.buffer[self.head] = value
        if self.head == self.tail and self.lastWasAdd:
            self.head = (self.head + 1) % len(self.buffer)
            self.tail = self.tail
        else:
            self.head = (self.head + 1) % len(self.buffer)
        self.lastWasAdd = True

    def remove(self, count):
        for _ in xrange(count):
            self.tail = (self.tail + 1) % len(self.buffer)
            if self.tail == self.head:
                break
        self.lastWasAdd = False

    def list(self):
        if self.empty():
            return
        pos = self.tail
        while True:
            print self.buffer[pos]
            pos = (pos + 1) % len(self.buffer)
            if pos == self.head:
                break

    def empty(self):
        return self.head == self.tail and not lastWasAdd

# read size of buffer from stdin
size = int(sys.stdin.readline())

buffer = circular_buffer(size)

# read command from stdin
while True:
    command = sys.stdin.read(1)

    if command == 'A':
        count = int(sys.stdin.readline())
        for _ in xrange(count):
            line = sys.stdin.readline().rstrip('\n')
            buffer.add(line)
    elif command == 'R':
        count = int(sys.stdin.readline())
        buffer.remove(count)
    elif command == 'L':
        buffer.list()
        sys.stdin.readline() # discard the newline
    elif command == 'Q':
        sys.exit(0)
    else:
        sys.exit(1)

File: ruby/question1/question1.rb
#!/usr/bin/env ruby

class CircularBuffer
  def initialize(size)
    @buffer = Array.new(size)
    @head = 0
    @tail = 0
    @lastWasAdd = false
  end

  def add(value)
    @buffer[@head] = value
    if @head == @tail && @lastWasAdd
      @head = (@head + 1) % @buffer.length
      @tail = @head
    else
      @head = (@head + 1) % @buffer.length
    end
    @lastWasAdd = true
  end

  def remove(count)
    return if empty?
    count.times do
      @tail = (@tail + 1) % @buffer.length
      break if @tail == @head
    end
    @lastWasAdd = false
  end

  def empty?
    @head == @tail && !@lastWasAdd
  end

  def list
    return if empty?
    pos = @tail
    loop do
      puts @buffer[pos]
      pos = (pos + 1) % @buffer.size
      break unless pos != @head
    end
  end
end

# read size of buffer
size = ARGF.readline.to_i

buffer = CircularBuffer.new(size)

loop do
  case ARGF.readchar
  when 'A'
    n = ARGF.readline.to_i
    # add n lines
    n.times do
      line = ARGF.readline
      buffer.add(line)
    end
  when 'R'
    n = ARGF.readline.to_i
    buffer.remove(n)
  when 'L'
    buffer.list
  when 'Q'
    exit 0
  else
    exit 1
  end
end

exit 2
